ntoskrnl/io/iomgr/driver.c

```diff
@@ -2181,4 +2194,233 @@ NtUnloadDriver(IN PUNICODE_STRING DriverServiceName)
+    HANDLE RootKeyHandle, KeyHandle, NodeHandle;
+    PKEY_VALUE_FULL_INFORMATION KeyValueInfo;
+    ULONG RegValue;
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    DPRINT("legacy driver: '%wZ'\n", ServiceName);
+
+    Status = IopOpenRegistryKeyEx(&RootKeyHandle,
+                                  NULL,
+                                  &RootKeyPath,
+                                  KEY_ALL_ACCESS);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    // create sub key and try to open or create it
+    SubKeyName.MaximumLength = LegacyPrefix.Length + ServiceName->Length + sizeof(UNICODE_NULL);
+    SubKeyName.Length = 0;
+    SubKeyName.Buffer = ExAllocatePoolWithTag(PagedPool, SubKeyName.MaximumLength, TAG_IO);
+    if (SubKeyName.Buffer == NULL)
+    {
+        DPRINT1("No SubKeyName.Buffer!\n");
+        ZwClose(RootKeyHandle);
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    RtlAppendUnicodeStringToString(&SubKeyName, &LegacyPrefix);
+    RtlAppendUnicodeStringToString(&SubKeyName, ServiceName);
+
+    Status = IopCreateRegistryKeyEx(&KeyHandle,
+                                    RootKeyHandle,
+                                    &SubKeyName,
+                                    KEY_ALL_ACCESS,
+                                    REG_OPTION_NON_VOLATILE,
+                                    NULL);
+
+    ExFreePool(SubKeyName.Buffer);
+    ZwClose(RootKeyHandle);
+
+    if (!NT_SUCCESS(Status))
+    {
+        ZwClose(KeyHandle);
+        return Status;
+    }
+
+    RtlInitUnicodeString(&KeyString, L"NextInstance");
+    RegValue = 1;
+    Status = ZwSetValueKey(KeyHandle,
+                           &KeyString,
+                           0,
+                           REG_DWORD,
+                           &RegValue,
+                           sizeof(RegValue));
+    if (!NT_SUCCESS(Status))
+    {
+        ZwClose(KeyHandle);
+        return Status;
+    }
+
+    RtlInitUnicodeString(&KeyString, L"000");
+    Status = IopCreateRegistryKeyEx(&NodeHandle,
+                                    KeyHandle,
+                                    &KeyString,
+                                    KEY_ALL_ACCESS,
+                                    REG_OPTION_NON_VOLATILE,
+                                    NULL);
+
+    ZwClose(KeyHandle);
+
+    if (!NT_SUCCESS(Status))
+    {
+        return Status;
+    }
+
+    Status = RtlDuplicateUnicodeString(RTL_DUPLICATE_UNICODE_STRING_NULL_TERMINATE,
+                                       ServiceName,
+                                       &ServiceName1);
+    if (!NT_SUCCESS(Status))
+    {
+        return Status;
+    }
+
+    RtlInitUnicodeString(&KeyString, L"Service");
+    Status = ZwSetValueKey(NodeHandle,
+                           &KeyString,
+                           0,
+                           REG_SZ,
+                           ServiceName->Buffer,
+                           ServiceName->Length + sizeof(UNICODE_NULL));
```

Comment:
Forgot to change these back:
```suggestion
                           ServiceName1.Buffer,
                           ServiceName1.Length + sizeof(UNICODE_NULL));
```